frontend/ui/trapper.lua

```diff
@@ -0,0 +1,337 @@
+--[[--
+Trapper module: provides methods for simple interaction with UI, to be
+used by linear jobs between their steps.
+
+Allows code to trap UI (give progress info to UI, ask for user choice),
+or get trapped by UI (get interrupted).
+Mostly done with coroutines, but hides their usage for simplicity.
+]]
+
+
+local ConfirmBox = require("ui/widget/confirmbox")
+local InfoMessage = require("ui/widget/infomessage")
+local UIManager = require("ui/uimanager")
+local ffiutil = require("ffi/util")
+local logger = require("logger")
+local _ = require("gettext")
+
+local Trapper = {}
+
+--[[--
+Execute function and allows it to be trapped (that is, to use our other methods)
+
+Simple wrapper function for a coroutine, which is a prerequisite
+for all our methods (this simply abstracts the @{coroutine}
+business to our callers), and execute it.
+This call should be the last step in some event processing code,
+as it may return early (the first @{coroutine.yield|coroutine.yield()} in any of the other
+methods will return from this function), and later be resumed by @{ui.uimanager|UIManager}.
+So any following (unwrapped) code would be then executed while `func`
+is half-done, with unintended consequences.
+
+@param func function reference to function to wrap and execute
+]]
+function Trapper:wrap(func)
+    -- Catch and log any error happening in func (an error happening
+    -- in a coroutine just aborts silently the coroutine)
+    local func = function()
+        ok, err = pcall(func)
+        if not ok then
+            logger.warn("error in wrapped function:", err)
+            return false
+        end
+        return true
+        -- As a coroutine, we will return at first coroutine.yield(),
+        -- and the above true/false won't probably be caught by
+        -- any code, but let's do it anyway.
+    end
+    local co = coroutine.create(func)
+    return coroutine.resume(co)
+end
+
+--- Is the code wrapped by Trapper?
+-- @treturn bool true if code is wrapped by Trapper, false otherwise
+function Trapper:isWrapped()
+    if coroutine.running() then
+        return true
+    end
+    return false
+end
+
+--- Allow for overriding the Paused text and buttons on the abort
+-- ConfirmBox displayed when InfoMessage dismissed in Trapper:UIFeedback()
+function Trapper:setPauseText(text, ok_text, cancel_text)
+    self.paused_text = text
+    self.paused_continue_text = ok_text
+    self.paused_abort_text = cancel_text
+end
+
+--- Simple UI interaction via InfoMessage and ConfirmBox
+--
+-- Usage:
```

Comment:
For this use

```lua
@usage
    -- four space indent
    blabla
    -etc
```

And it should go at the bottom.

To test what the full docs will look like you can run `ldoc .` in /docs.

Or I'll play around with it a bit later like I did for the others. :-)

Reply:
Well, my debian ldoc, while parsing the whole tree, does not want to output anything. It does thus when i tell it to process one single lua file. And it does not produce a doc as nice as yours :|
Anyway, before documenting some possibly obsoleted function :) , tell me what you think about the changes proposed above (if we have multiple smaller methods, no need for `@usage` documenting a complicated api.

Reply:
You'd have to stick a config.ld in the dir where you run ldoc with `format = 'markdown'` for the formatting, but references won't work unless you run it on everything (which probably takes less than a second so there's hardly any reason not to).

From luarocks
```
ldoc, a documentation generator for Lua, vs 1.4.6
```

```
$ apt show lua-ldoc
Package: lua-ldoc
Version: 1.4.3-5+nmu1
```
I don't think it's changed all that much though?

Reply:
I have the same version as yours.
```
/koreader/doc $ ldoc .
reading configuration from config.ld
/koreader/frontend/apps/cloudstorage/cloudstorage.lua:376: no module() call found; no initial doc comment
[...]
/koreader/frontend/ui/screensaver.lua:154: no module() call found; no initial doc comment
/koreader/frontend/ui/trapper.lua:105: ?: unknown tag: 'text' nil
/koreader/frontend/ui/trapper.lua:105: ?: unknown tag: 'confirmbox' nil
/koreader/frontend/ui/trapper.lua:105: ?: unknown tag: 'ok' nil
/koreader/frontend/ui/trapper.lua:105: ?: unknown tag: 'cancel' nil
/koreader/frontend/ui/widget/bboxwidget.lua:1: contains no items
[...]
/koreader/base/ffi/pic.lua:251: ?: Tag without value: name
/koreader/doc $
```
No output produced. But this one produces some output:
```
/koreader/doc $ ldoc -c config.ld ../frontend
[... same output as above, plus:]
output written to /koreader/doc/doc
/koreader/doc $
```
although the HTML does not have any of the project/description... of the config.ld...
The internal links and links to lua documentation are there, but the `cmd` are there as is with their backticks, no monospace. I had installed lua-markdown 0.32-5 in case it matters.

Reply:
Odd, it's supposed to end with `output written to /home/frans/src/kobo/koreader/doc/html`

Regardless of Markdown, it's just nice to have all of the module references.